sunpy/coordinates/transformations.py

```diff
@@ -76,108 +63,155 @@ def _observers_are_equal(obs_1, obs_2, string_ok=False):
 
-    return heliopframe.realize_frame(representation)
 
-
-@frame_transform_graph.transform(FunctionTransform, Helioprojective,
-                                 Heliocentric)
+@frame_transform_graph.transform(FunctionTransformWithFiniteDifference,
+                                 Helioprojective, Heliocentric)
 def hpc_to_hcc(heliopcoord, heliocframe):
     """
     Convert from Helioprojective Cartesian to Heliocentric Cartesian.
     """
-    if not _observers_are_equal(heliopcoord.observer, heliocframe.observer):
-        heliocframe_heliopobs = Heliocentric(observer=heliopcoord.observer)
-        helioccoord_heliopobs = heliopcoord.transform_to(heliocframe_heliopobs)
-        helioccoord = helioccoord_heliopobs.transform_to(heliocframe)
-        return helioccoord
-
     if not isinstance(heliopcoord.observer, BaseCoordinateFrame):
         raise ConvertError("Cannot transform helioprojective coordinates to "
                            "heliocentric coordinates for observer '{}' "
                            "without `obstime` being specified.".format(heliopcoord.observer))
 
     heliopcoord = heliopcoord.calculate_distance()
-    x = np.deg2rad(heliopcoord.Tx)
-    y = np.deg2rad(heliopcoord.Ty)
 
-    cosx = np.cos(x)
-    sinx = np.sin(x)
-    cosy = np.cos(y)
-    siny = np.sin(y)
+    # Permute/swap axes from HPC equivalent Cartesian to HCC
+    newrepr = heliopcoord.cartesian.transform(matrix_transpose(_matrix_hcc_to_hpc()))
+
+    # Shift the origin from the observer to the Sun
+    distance = SkyCoord(heliocframe.observer).heliographic_stonyhurst.radius
+    newrepr += CartesianRepresentation(0*u.m, 0*u.m, distance)
+
+    # Complete the conversion of HPC to HCC at the obstime and observer of the HPC coord
+    int_coord = Heliocentric(newrepr, obstime=heliopcoord.obstime, observer=heliopcoord.observer)
+
+    # Loopback transform HCC as needed
+    return int_coord.transform_to(heliocframe)
+
+
+def _rotation_matrix_hcc_to_hgs(longitude, latitude):
+    # Returns the rotation matrix from HCC to HGS based on the observer longitude and latitude
 
-    rx = (heliopcoord.distance.to(u.m)) * cosy * sinx
-    ry = (heliopcoord.distance.to(u.m)) * siny
-    rz = (heliopcoord.observer.radius.to(u.m)) - (
-        heliopcoord.distance.to(u.m)) * cosy * cosx
+    # Permute the axes of HCC to match HGS Cartesian equivalent
+    #   HGS_X = HCC_Z
+    #   HGS_Y = HCC_X
+    #   HGS_Z = HCC_Y
+    axes_matrix = np.array([[0, 0, 1],
+                            [1, 0, 0],
+                            [0, 1, 0]])
 
-    representation = CartesianRepresentation(
-        rx.to(u.km), ry.to(u.km), rz.to(u.km))
-    return heliocframe.realize_frame(representation)
+    # Rotate in latitude and longitude (sign difference because of direction difference)
+    lat_matrix = rotation_matrix(latitude, 'y')
+    lon_matrix = rotation_matrix(-longitude, 'z')
 
+    return matrix_product(matrix_product(lon_matrix, lat_matrix), axes_matrix)
```

Comment:
The future is now:
```suggestion
    return lon_matrix @ lat_matrix @ axes_matrix
```
(Also the docstring for `matrix_product` tells me it exists so you don't need to nest calls to it.)

Reply:
Really?  Sweet!